justfile

```diff
@@ -0,0 +1,156 @@
+
+# Make sure that some env variables are set for all jobs.
+#FIXME: Consider using .env for this, but only if used also on CI
+export RUST_WORKSPACE := env_var_or_default("RUST_WORKSPACE", "discovery_engine_core")
+export DART_WORKSPACE := env_var_or_default("DART_WORKSPACE", "discovery_engine")
+export CARGO_INSTALL_ROOT := env_var_or_default("CARGO_INSTALL_ROOT", "cargo-installs")
+
+# Run just --list
+default:
+    @{{just_executable()}} --list
+
+# Get/update dart deps
+dart_deps:
+    cd "$DART_WORKSPACE"; \
+    dart pub get
+
+# Fetch rust dependencies
+rust_deps:
+    cd "$RUST_WORKSPACE"; \
+    cargo fetch --locked
+
+# Installs async-bindgen cli tool
+install_async_bindgen *args:
+    cargo install \
+        --git https://github.com/xaynetwork/xayn_async_bindgen.git \
+        {{args}} \
+        async-bindgen-gen-dart \
+        "$@" ;
+
+# Get/Update/Fetch/Install all dependencies
+deps: dart_deps rust_deps install_async_bindgen
+
+# Format dart (checks only on CI)
+dart_fmt:
+    cd "$DART_WORKSPACE"; \
+    dart format {{ if env_var_or_default("CI", "false") == "true" { "--output=none --set-exit-if-changed" } else { "" } }} .
+
+# Format rust (checks only on CI)
+rust_fmt:
+    cd "$RUST_WORKSPACE"; \
+    cargo +nightly fmt --all -- {{ if env_var_or_default("CI", "false") == "true" { "--check" } else { "" } }};\
+    cargo sort --grouped --workspace {{ if env_var_or_default("CI", "false") == "true" { "--check" } else { "" } }}
+
+# Format all code (checks only on CI)
+fmt: rust_fmt dart_fmt
+
+# Check dart code, fails on info on CI
+dart_check: dart_build
+    cd "$DART_WORKSPACE"; \
+    dart analyze {{ if env_var_or_default("CI", "false") == "true" { "--fatal-infos" } else { "" } }}
+
+_codegen_order_workaround:
```

Comment:
To me it's not obvious what this does or why it's needed. Since it's labeled workaround, could you add a comment explaining why this is needed?

Reply:
We use cbindgen to generate C headers for ffi functions.

We use async_bindgen to generate some of the C functions.

The problem is that cbindgen doesn't the code generated by derives and pro-macros, normal macros etc. (because it doesn't tightly integrate into rust, in turn you can use it without running cargo).

So it doesn't see our generate C functions.

As a ad-hoc fix for this we wrote the generated functions into a file.

The problem is that the file is generate after cbindgen runs, also after rust parsed all files.

Hence we must run a (likely failing) cargo check before running what we actually want. A bit like in latex.

This obvious is not the best solution to the problem, just the fastest to get working.

But replacing it with a more proper solution has currently no-priority.

Ironically I missed this problem for a while because vscode being so nice as to doing that "pre-run" for me, sometimes even after I closed it to test exactly that case (there is/was some bug where in some edge cases rust-analyzer continues to run for a while even after you close the workspace).

Reply:
Thanks for the in-depth explanation :) Could you also put a condensed version in the justfile as a comment? :>